### ISSUE TITLE
Add portals (set/get variable) to the visual shader editor

### ISSUE THREAD

**Issue by SubtleMetaphor** [godotengine/godot-proposals#7857]
### Describe the project you are working on

Godot

### Describe the problem or limitation you are having in your project

When working with the Shader Editor (creating shaders visually using nodes) it can become quite tangled or difficult to organize.

Here's an example from Unity's shader editor (which does not support portals/variables), where things have gotten frustratingly entangled:

![82a0b4130e78a4af1511df02bb0e72fa](https://github.com/godotengine/godot-proposals/assets/6630276/d2553802-458b-455f-a86b-092dd1ab7a6b)

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

One feature that some tools or engines have is the ability to create a "portal" inside the shader editor. Essentially a node for setting a variable, and later getting it at some stage. It is, in a way, similar to the varyings implementation in Godot. However, varyings is a way for the vertex shader to communicate with the fragment shader, and that is not what this is. There may be portals that only exist within the vertex shader, or only within the fragment shader. Using a varying for this is not ideal for a lot of situations where you want higher fidelity (by only using fragment shader), or where you only want the data to exist within the vertex shader.

![61559c3289aad2a5af18074e0b815328](https://github.com/godotengine/godot-proposals/assets/6630276/9f4d690d-3175-4ee0-b0bd-f0224cb445ee)
_(Screenshot of a shader made in Amplify Shader Editor for Unity, which does support setting/getting variables.)_

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Here's close up shots of portals/variable set/get.

![386d30599d2e0fdc7f16c346936f0ece](https://github.com/godotengine/godot-proposals/assets/6630276/e023e0ca-1e44-42c5-9955-115b6041e2d0)
_(From this screenshot in Amplify Shader Editor, vertexBitangent is set, and then used in the "line" below it.)_

![495ca19baf09442980ed779d7af661bb](https://github.com/godotengine/godot-proposals/assets/6630276/34093f99-7e08-4d48-aea4-1d464bd0089d)
_(Here's a screenshot from Unigine's shader editor, which does support portals.)_

### If this enhancement will not be used often, can it be worked around with a few lines of script?

N/A

### Is there a reason why this should be core and not an add-on in the asset library?

N/A

**Comment by Chaosus** [comment]
Isn't it a duplicate of https://github.com/godotengine/godot-proposals/issues/7730 (same functionality)

**Comment by SubtleMetaphor** [comment]
> Isn't it a duplicate of #7730 ? (same functionality)

No. That suggestion is like defining a function. My request is more like setting a variable and then referencing it.

**Comment by QbieShay** [comment]
This should be done with the ref node I think? @Chaosus ?
